module/TestModule/src/TestModule.cpp

```diff
@@ -0,0 +1,290 @@
+            cfg.send_port    = config["send_port"].as<uint>();
+            cfg.recieve_port = config["receive_port"].as<uint>();
+        });
+
+        on<Every<5, Per<std::chrono::seconds>>,
+           With<GlobalConfig>,
+           With<Sensors>,
+           With<Field>,
+           With<LocalisationBall>,
+           With<KickPlan>,
+           With<GameState>,
+           With<WalkCommand>,
+           Single>()
+            .then([this](const std::shared_ptr<const GlobalConfig>& global,
+                         const std::shared_ptr<const Sensors>& sensors,
+                         const std::shared_ptr<const Field>& field,
+                         const std::shared_ptr<const LocalisationBall>& loc_ball,
+                         const std::shared_ptr<const KickPlan>& kick_plan,
+                         const std::shared_ptr<const GameState>& game_state,
+                         const std::shared_ptr<const WalkCommand>& walk_command) {
+                auto packet = std::make_unique<Message>();
+
+                // 1. timestamp
+                packet->timestamp = NUClear::clock::now();
+
+                auto pr = game_state->data.self.penalty_reason;
+
+                // 2. game state
+                if (pr == PenaltyReason::UNPENALISED) {
+                    packet->state = State::UNPENALISED;
+                }
+                else if (pr >= PenaltyReason::UNKNOWN_PENALTY_REASON && pr <= PenaltyReason::PLAYER_PUSHING) {
+                    packet->state = State::PENALISED;
+                }
+                else {
+                    packet->state = State::UNKNOWN_STATE;
+                }
+
+                // 3. current pose
+                // 3.1 player id
+
+                packet->current_pose.player_id = global->player_id;
+
+                // transformations to field space
+                Eigen::Isometry3d Htw(sensors->Htw);
+
+                Eigen::Isometry3d Hfw;
+                Eigen::Isometry2d position(field->position);
+
+                Hfw.translation() = Eigen::Vector3d(position.translation().x(), position.translation().y(), 0);
+                Hfw.linear() =
+                    Eigen::AngleAxisd(Eigen::Rotation2Dd(position.rotation()).angle(), Eigen::Vector3d::UnitZ())
+                        .toRotationMatrix();
+
+                Eigen::Isometry3d Hft = Hfw * Htw.inverse();
+                Eigen::Vector3d rTFf  = Hft.translation();
+
+                // 3.2 position and covariance
+                packet->current_pose.position =
+                    Eigen::Vector3f(rTFf.x(), rTFf.y(), Hft.rotation().matrix().eulerAngles(0, 1, 2).z());
+                packet->current_pose.covariance = field->covariance.cast<float>();
+
+                // 4. walk command
+                // TODO x and y velocity estimations
+
+                Eigen::Vector3f motion_walk_command = walk_command->command.cast<float>();
+                packet->walk_command.z()            = motion_walk_command.z();
```

Comment:
definitely could do, the reason I split it up like that is because WalkCommand is defined as (x position, y position and rotational velocity). The Robocup proto puts all the velocities in walk_command, and all the robot positions in target_pose so I split up the WalkCommand proto data between the two